src/HigherOrderExpectation.php

```diff
@@ -26,57 +28,39 @@ final class HigherOrderExpectation
     private $expectation;
 
     /**
-     * @var bool
+     * @var mixed
      */
-    private $opposite = false;
+    private $value;
 
     /**
-     * @var string
+     * @var mixed
      */
-    private $name;
+    private $initialValue;
 
     /**
-     * Creates a new higher order expectation.
-     *
-     * @param array<int|string, mixed>|null $parameters
-     * @phpstan-ignore-next-line
+     * @var bool
      */
-    public function __construct(Expectation $original, string $name, ?array $parameters = null)
-    {
-        $this->original = $original;
-        $this->name     = $name;
-
-        $this->expectation = $this->expect(
-            is_null($parameters) ? $this->getPropertyValue() : $this->getMethodValue($parameters)
-        );
-    }
+    private $opposite = false;
 
     /**
-     * Retrieves the property value from the original expectation.
-     *
-     * @return mixed
+     * @var bool
      */
-    private function getPropertyValue()
-    {
-        if (is_array($this->original->value)) {
-            return $this->original->value[$this->name];
-        }
-
-        // @phpstan-ignore-next-line
-        return $this->original->value->{$this->name};
-    }
+    private $lastCallWasAssertion = false;
 
     /**
-     * Retrieves the value of the method from the original expectation.
+     * Creates a new higher order expectation.
      *
-     * @param array<int|string, mixed> $arguments
+     * @param mixed $value
+     * @param mixed $initialValue
      *
-     * @return mixed
+     * @phpstan-ignore-next-line
      */
-    private function getMethodValue(array $arguments)
+    public function __construct(Expectation $original, $value, $initialValue = null)
     {
-        // @phpstan-ignore-next-line
-        return $this->original->value->{$this->name}(...$arguments);
+        $this->original     = $original;
+        $this->expectation  = $this->expect($value);
+        $this->value        = $value;
+        $this->initialValue = $initialValue ?? $original->value;
```

Comment:
Is this property needed, will it only ever be the original->value anyway?

Reply:
🔥